sql/src/main/java/io/crate/planner/operators/LogicalPlan.java

```diff
@@ -100,6 +101,15 @@
          *                       Collect (usedColumns: [a] - inherited from Order)
          *                         outputs: [_fetch, a]
          *                    </pre>
+         *
+         *                    These plans might later be optimized, see #{@link LogicalPlan#tryOptimize(LogicalPlan)},
+         *                    and as such some operators need to identify columns in the "parent" that occur in this
+         *                    "child".
+         *                    Some implementations might use an index based match (ie. column number 1 in the parent's
```

Comment:
Which implementations do that? Shouldn't we change the data structure to something order preserving, e.g. List?

Reply:
It's the Order implementation, I didn't mention it as it's well, an implementation. Think we should? (to the client/user of this interface it shouldn't matter)

I wouldn't want to change it to a List as part of an impediment as that might open a can of worms (since the Sets are duplicate free already)

Reply:
Just asking because I wasn't sure which operators depend on it. I agree changing might be out of scope, just a suggestion. I wonder, have all operators been updated? Union for instance still uses HashSet.

Reply:
Could we mark this with like a big TODO FIX ME. Wouldn't it suffice to change the `usedBeforeNextFetch` to type `LinkedHashSet`? That would catch all possible regressions.

Reply:
I wonder if we should go the other way around and change the logic so that the deterministic order isn't required anymore?

Reply:
+1 That would get to the root of the problem instead of curing symptoms.

Reply:
@mfussenegger that is also an option, we thought that requires more work and shouldn't be an impediment. @matriv is currently researching a way to maybe drop it as we have to work on the push down optimisation for the HashJoin (where it should be applied on the right branch as oppose to the left like we currently do for the NL Join). 
Hm, maybe it does make sense to revert this fix on master, keep using HashSet and see what we find out as part of Marios's research?

Reply:
For Union we would always require the positional replacement of orderBy symbols, so I don't see a way to avoid the LinkedHashSet to preserve the ordering.

Reply:
@mfussenegger @mxm @matriv Actually, the optimization refactoring for hash join is optional for our story so we should go ahead with this fix and if we get to redesign how we do the push down optimisation it'll be a bonus for this sprint (but not sure we'll get to it)

Reply:
>For Union we would always require the positional replacement of orderBy symbols, so I don't see a way to avoid the LinkedHashSet to preserve the ordering.

That's why I asked because we currently doesn't use LinkedHashSet but HashSet in Union. Just making sure we replaced it everywhere.

Reply:
Didn't want to block this but we should definitely keep track of this for the cleanup phase or so.

Reply:
Made the suggested changes - see https://github.com/crate/crate/pull/6921/commits/1b864309abe3e3c26a6ba8290f304b62a39fa07b and fixup